### PR TITLE
Forward backward correlate

### PR DESCRIPTION
**Pull request by gaoteng-git**
Add new link event "forward_backward" to dumped json file. 
The latency overhead is only about 1% to `profile.__exit__`

image
image

### CODE REVIEW

**Comment by chaekit** [comment]
@gdankel +1 on adding extra fields in GTA after all the refactoring we have gone through. I'll add a comment in the class to warn future contributor about this. other than that @gaoteng-git thanks for the PR!

File: libkineto/include/GenericTraceActivity.h
Comment (gdankel) [inline libkineto/include/GenericTraceActivity.h:90]: Adding four fields for this is a bit much. And it's very specific, both to PyTorch and to the way forward / backwards currently works.
At the very least this needs a big FIXME comment to explain this is a temporary and hacky implementation.

File: libkineto/src/ActivityProfiler.cpp
Comment (gdankel) [inline libkineto/src/ActivityProfiler.cpp:215]: When does it happen that the same key is encountered multiple times?

**Review by gdankel — COMMENTED** [review]
I think this is good for a prototype implementation but we can't land this without big comments in the code referring to an issue describing a plan for improving.

File: libkineto/src/ActivityProfiler.cpp
Comment (gaoteng-git) [reply libkineto/src/ActivityProfiler.cpp:215]: Now the sequence number is only incremented on creating a "Node" object for backward pass, by [calling "at::sequence_number::get_and_increment()"](https://github.com/pytorch/pytorch/blob/bf7d03ff1f3cded53a16b707a2db76ce25c130f7/torch/csrc/autograd/function.h#L130). 
Among all ops with same sequence number, the one with biggest startTime is the one launching backward op. So I choose this one to cache into "tidSeq2activity_" for matching with backward-op of same tid and seq number.
I've added comments explaining this.

File: libkineto/include/GenericTraceActivity.h
Comment (gaoteng-git) [reply libkineto/include/GenericTraceActivity.h:90]: Yes, it's a little heavy. I've considered moving these columns into "args" json string, but it will need parse the string in kineto side when using them to do correlation.
I've added FIXME to it. Please feel free to update it in a better way later on.

Comment (gdankel) [reply libkineto/include/GenericTraceActivity.h:91]: I think most of the code does not belong in Kineto itself but in PyTorch Kineto profiler.

File: libkineto/src/ActivityProfiler.cpp
Comment (gdankel) [inline libkineto/src/ActivityProfiler.cpp:194]: I suggest adding all this code into profiler_kineto.cpp (as part of #62553) and call it from finalizeCPUTrace.
Then, the only thing we need to add to GenericTraceActivity is a linkedFlowId or a pointer to the linked activity that can be returned by linkedActivity().
My overall plan for flows is to encode a flow ID and a flow type in a flow field (e.g. 4 bits for flow type). Profilers can register flow types and associate them with names to make it possible to link activities across separate profilers.
In this case you're only linking activities within a single profiler so it's a bit simpler.

File: libkineto/src/output_json.cpp
Comment (gdankel) [inline libkineto/src/output_json.cpp:248]: Change to check for linkedActivity() != nullptr after making changes suggested above, and add handleGenericLink() instead taking two activities (start and end).

**Review by gdankel — CHANGES_REQUESTED** [review]

File: libkineto/include/GenericTraceActivity.h
Comment (gaoteng-git) [reply libkineto/include/GenericTraceActivity.h]: OK. Moved to PyTorch side.

File: libkineto/src/ActivityProfiler.cpp
Comment (gaoteng-git) [reply libkineto/src/ActivityProfiler.cpp]: Got it. Made following change to "GenericTraceActivity.h":
Added "TraceActivity* linkedAct;" to point to its corresponding backward or forward op. Added "uint64_t linkId;" to give a pair of forward&backward ops same value, the high-4bits is link type.
And move correlating logic to "finalizeCPUTrace".

File: libkineto/src/output_json.cpp
Comment (gaoteng-git) [reply libkineto/src/output_json.cpp]: Thanks. Done.

File: libkineto/src/output_json.h
Comment (gdankel) [inline libkineto/src/output_json.h:47]: Move these to private

Comment (gaoteng-git) [reply libkineto/src/output_json.h]: Thanks. Done.

File: libkineto/include/GenericTraceActivity.h
Comment (gdankel) [inline libkineto/include/GenericTraceActivity.h:100]: From offline discussion, how about this for now:
struct {
TraceActivity* prevActivity;
uint64_t id : 28;
uint64_t type : 4;
} flow;

prevActivity establishes order, so even if timestamps are equal we know which come first.

File: libkineto/include/GenericTraceActivity.h
Comment (gdankel) [inline libkineto/include/GenericTraceActivity.h:19]: I think we can remove all these defines with the below bitfield proposal.

File: libkineto/include/GenericTraceActivity.h
Comment (gdankel) [inline libkineto/include/GenericTraceActivity.h:25]: We can use one type here I think, just to determine category and name:
constexpr unsigned int kLinkFwdBwd = 1;

File: libkineto/src/output_json.cpp
Comment (gdankel) [inline libkineto/src/output_json.cpp:255]: Assuming linkedActivity is set only for destination and above bitfield suggestion:

if (op.flow.linkType == kLinkFwdBwd) {
  handleGenericLinkStart(op.flow.linkedActivity(), "forward_backward", "fwd_bwd");
  handleGenericLinkEnd(op, "forward_backward", "fwd_bwd");
}

File: libkineto/src/output_json.cpp
Comment (gdankel) [inline libkineto/src/output_json.cpp:274]: s.flow.id

**Review by gdankel — CHANGES_REQUESTED** [review]

File: libkineto/include/GenericTraceActivity.h
Comment (gdankel) [inline libkineto/include/GenericTraceActivity.h:33]: linkId?

File: libkineto/src/output_json.cpp
Comment (gdankel) [inline libkineto/src/output_json.cpp:260]: rename to handleGenericLinkStart()?

File: libkineto/src/output_json.cpp
Comment (gdankel) [inline libkineto/src/output_json.cpp:278]: rename?

**Comment by gdankel** [comment]
LGTM, just need to pass tests and address nits.

**Review by gdankel — APPROVED** [review]

File: libkineto/src/output_json.cpp
Comment (gaoteng-git) [reply libkineto/src/output_json.cpp]: Thanks. Done.

File: libkineto/src/output_json.cpp
Comment (gaoteng-git) [reply libkineto/src/output_json.cpp]: Thanks. Done.

File: libkineto/include/GenericTraceActivity.h
Comment (gaoteng-git) [reply libkineto/include/GenericTraceActivity.h]: Thanks. Done.

File: tb_plugin/torch_tb_profiler/profiler/multiprocessing.py
Comment (cloudhan) [reply tb_plugin/torch_tb_profiler/profiler/multiprocessing.py:6]: @chaekit Maybe we could do a cherry-pick to the master branch and remove the commit from this branch? What do you think?

Comment (cloudhan) [reply tb_plugin/torch_tb_profiler/profiler/multiprocessing.py:6]: That will create a merge conflict when we want to merge plugin/vnext to master but we can resolve it on our side.

Comment (cloudhan) [reply tb_plugin/torch_tb_profiler/profiler/multiprocessing.py:6]: @gaoteng-git Please consider merge this PR after https://github.com/pytorch/kineto/pull/402

**Review by chaekit — APPROVED** [review]

**Pull request closed without merging**
